ocaml/typing/jkind.ml

```diff
@@ -549,74 +566,103 @@ module Desc = struct
+    | Var _, Const c when Const.equal Const.max c -> Less
     | Var v1, Var v2 -> if v1 == v2 then Equal else Not_le
     | Const _, Var _ | Var _, Const _ -> Not_le
 end
 
 module Jkind_desc = struct
   type t =
     { layout : Layout.t;
+      modes : Alloc.Const.t;
       externality : Externality.t
     }
 
-  let max = { layout = Layout.max; externality = Externality.max }
+  let max =
+    { layout = Layout.max;
+      modes = Alloc.Const.max;
+      externality = Externality.max
+    }
 
-  let equate_or_equal ~allow_mutation { layout = lay1; externality = ext1 }
-      { layout = lay2; externality = ext2 } =
+  let equate_or_equal ~allow_mutation
+      { layout = lay1; modes = modes1; externality = ext1 }
+      { layout = lay2; modes = modes2; externality = ext2 } =
     Layout.equate_or_equal ~allow_mutation lay1 lay2
+    && Alloc.Const.equal modes1 modes2
     && Externality.equal ext1 ext2
 
-  let sub { layout = layout1; externality = externality1 }
-      { layout = layout2; externality = externality2 } =
-    Misc.Le_result.combine
-      (Layout.sub layout1 layout2)
-      (Externality.less_or_equal externality1 externality2)
+  let sub { layout = lay1; modes = modes1; externality = ext1 }
+      { layout = lay2; modes = modes2; externality = ext2 } =
+    Misc.Le_result.combine_list
+      [ Layout.sub lay1 lay2;
+        Alloc.Const.less_or_equal modes1 modes2;
+        Externality.less_or_equal ext1 ext2 ]
 
-  let intersection { layout = lay1; externality = ext1 }
-      { layout = lay2; externality = ext2 } =
+  let intersection { layout = lay1; modes = modes1; externality = ext1 }
+      { layout = lay2; modes = modes2; externality = ext2 } =
     Option.bind (Layout.intersection lay1 lay2) (fun layout ->
-        Some { layout; externality = Externality.meet ext1 ext2 })
+        Some
+          { layout;
+            modes = Alloc.Const.meet modes1 modes2;
+            externality = Externality.meet ext1 ext2
+          })
 
   let of_new_sort_var () =
     let layout, sort = Layout.of_new_sort_var () in
-    { layout; externality = Externality.max }, sort
+    { max with layout }, sort
 
   let any = max
 
-  let value = { layout = Layout.value; externality = Externality.max }
+  let value = { max with layout = Layout.value }
 
-  let void = { layout = Layout.void; externality = Externality.min }
+  let void =
+    { layout = Layout.void;
+      modes = Alloc.Const.max;
+      externality = Externality.min
+    }
 
-  let immediate64 = { layout = Layout.value; externality = External64 }
+  let immediate64 =
+    { layout = Layout.value;
+      modes = { locality = Global; linearity = Many; uniqueness = Unique };
+      externality = External64
+    }
 
-  let immediate = { layout = Layout.value; externality = External }
+  let immediate =
+    { layout = Layout.value;
+      modes = { locality = Global; linearity = Many; uniqueness = Unique };
+      externality = External
+    }
```

Comment:
For the monadic axes (e.g. uniqueness), I'm not sure what's the best way to characterise them. A value being `unique` means "you can strong update it if it's a record" or "it has no other aliases"? Consider the example: 
```
type t = { x : Int63.t} [@@unboxed}
```
This is immediate only on 64-bit systems. If we allow it cross to `unique` even on 32-bit system, then we should disallow strong update on this type.

(Similar for the `contention` axis).

I have an idea which is probably silly: can we make kinds platform-dependent? That is, we will only have a single `immediate` kind, and writing `type t : immediate = Int63.t` will error on 32bit systems?

Reply:
I implicitly assumed (but actually did think about) that we couldn't in-place update unboxed records. In-place update is about reusing memory... but an unboxed record doesn't have memory to reuse, so I'm not exactly sure what it would mean.

_Can_ we make kinds platform-dependent? Sure. _Should_ we make kinds platform-dependent? Almost certainly not. We don't want our type system to not be portable across machines. That's a recipe for devops sadness.

I agree there may be trouble in this area someday. But that day is not today!

Reply:
I'm interested in extending the conversation about platform-dependent kinds. But let's not block this PR.